pageserver/src/page_cache.rs

```diff
@@ -226,51 +254,290 @@ pub struct CacheEntryContent {
         CacheEntry {
             key,
-            content: Mutex::new(CacheEntryContent {
-                page_image: None,
-                wal_record: None,
-                apply_pending: false,
-            }),
+            content: Mutex::new(content),
             walredo_condvar: Condvar::new(),
         }
     }
 }
 
-#[derive(Eq, PartialEq, Hash, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Hash, Ord, Clone, Copy)]
 pub struct RelTag {
     pub spcnode: u32,
     pub dbnode: u32,
     pub relnode: u32,
     pub forknum: u8,
 }
 
-#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
+impl RelTag {
+    pub fn pack(&self, buf: &mut BytesMut) {
+        buf.put_u32(self.spcnode);
+        buf.put_u32(self.dbnode);
+        buf.put_u32(self.relnode);
+        buf.put_u32(self.forknum as u32);
+    }
+    pub fn unpack(buf: &mut BytesMut) -> RelTag {
+        RelTag {
+            spcnode: buf.get_u32(),
+            dbnode: buf.get_u32(),
+            relnode: buf.get_u32(),
+            forknum: buf.get_u32() as u8,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
 pub struct BufferTag {
-    pub spcnode: u32,
-    pub dbnode: u32,
-    pub relnode: u32,
-    pub forknum: u8,
+    pub rel: RelTag,
     pub blknum: u32,
 }
 
+impl BufferTag {
+    pub fn pack(&self, buf: &mut BytesMut) {
+        self.rel.pack(buf);
+        buf.put_u32(self.blknum);
+    }
+    pub fn unpack(buf: &mut BytesMut) -> BufferTag {
+        BufferTag {
+            rel: RelTag::unpack(buf),
+            blknum: buf.get_u32(),
+        }
+    }
+}
+
 #[derive(Clone)]
 pub struct WALRecord {
     pub lsn: u64, // LSN at the *end* of the record
     pub will_init: bool,
+    pub truncate: bool,
     pub rec: Bytes,
     // Remember the offset of main_data in rec,
     // so that we don't have to parse the record again.
     // If record has no main_data, this offset equals rec.len().
-    pub main_data_offset: usize,
+    pub main_data_offset: u32,
+}
+
+impl WALRecord {
+    pub fn pack(&self, buf: &mut BytesMut) {
+        buf.put_u64(self.lsn);
+        buf.put_u8(self.will_init as u8);
+        buf.put_u8(self.truncate as u8);
+		buf.put_u32(self.main_data_offset);
+        buf.put_u32(self.rec.len() as u32);
+        buf.put_slice(&self.rec[..]);
+    }
+    pub fn unpack(buf: &mut BytesMut) -> WALRecord {
+        let lsn = buf.get_u64();
+        let will_init = buf.get_u8() != 0;
+        let truncate = buf.get_u8() != 0;
+		let main_data_offset = buf.get_u32();
+        let mut dst = vec![0u8; buf.get_u32() as usize];
+        buf.copy_to_slice(&mut dst);
+        WALRecord {
+            lsn,
+            will_init,
+            truncate,
+            rec: Bytes::from(dst),
+			main_data_offset
+        }
+    }
 }
 
 // Public interface functions
 
 impl PageCache {
+	fn do_gc(&self, conf: &PageServerConf) -> anyhow::Result<Bytes> {
```

Comment:
This function creeps to the right so far, it's hard to follow the logic. Can we break it up into smaller functions?

Reply:
Yeh, I also find this function too long.And it contains similar fragments of code which is better to extract in some functions. But unfortunately I didn't find nice way to do it, becasue of several mutable variables used and updated by this fragments: minbuf, maxbuf, minkey. maxkey. I do not want to move them to the object because them are used locally only in this thread. And I want to minimize buffer allocations. 

Let me think more about it. I just do not want to formally satisfy your requirement, for example by extracting loop body in separate functions.